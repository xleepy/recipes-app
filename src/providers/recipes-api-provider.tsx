import { createContext, ReactNode, useMemo, useContext } from 'react';
import { RecipesApi } from '../api/apis/RecipesApi';
import { Configuration } from '../api/runtime';

type RecipesProviderContextValue = RecipesApi | null;

const RecipesProviderContext = createContext<RecipesProviderContextValue>(null);

export const RecipesApiProvider = ({ children }: { children: ReactNode }) => {
  const api = useMemo(() => {
    return new RecipesApi(
      new Configuration({ apiKey: import.meta.env.VITE_APP_ID })
    );
  }, []);
  return (
    <RecipesProviderContext.Provider value={api}>
      {children}
    </RecipesProviderContext.Provider>
  );
};

export const useRecipesApi = () => {
  const context = useContext(RecipesProviderContext);
  if (!context) {
    throw new Error('Api not provided');
  }
  return context;
};
