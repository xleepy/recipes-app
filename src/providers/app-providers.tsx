import { RecipesApiProvider } from './recipes-api-provider';
import { HashRouter } from 'react-router-dom';
import { ReactNode } from 'react';

type AppProvidersProps = {
  children: ReactNode;
};

export const AppProviders = ({ children }: AppProvidersProps) => {
  return (
    <RecipesApiProvider>
      <HashRouter>{children}</HashRouter>
    </RecipesApiProvider>
  );
};
