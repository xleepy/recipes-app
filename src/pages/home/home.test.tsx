import { render, waitFor } from '@testing-library/react';
import { describe, expect, it, vi } from 'vitest';
import { Home } from '.';
import {
  GetRecipeInformation200Response,
  SearchRecipes200Response,
} from '../../api';
import { MemoryRouter } from 'react-router-dom';
import { RecipesApiProvider } from '../../providers/recipes-api-provider';

vi.mock('../../api/apis/RecipesApi', () => {
  const MockInstance = vi.fn(() => {
    return {
      searchRecipes: vi.fn(() => {
        return Promise.resolve<SearchRecipes200Response>({
          results: new Set([
            {
              title: 'test',
              protein: 'yes',
              imageType: 'png',
              image: 'none',
              id: 1,
              fat: 'test',
              carbs: 'test',
              calories: 1,
            },
          ]),
          totalResults: 1,
          offset: 1,
          number: 1,
        });
      }),
      getRecipeInformation: vi.fn(() => {
        return Promise.resolve<GetRecipeInformation200Response>({
          title: 'test',
          imageType: 'png',
          image: 'none',
          id: 1,
          servings: 1,
          vegan: true,
          whole30: false,
          weightWatcherSmartPoints: 1,
          veryPopular: false,
          veryHealthy: true,
          vegetarian: true,
          sustainable: true,
          summary: 'lorem ipsum',
          spoonacularScore: 1,
          spoonacularSourceUrl: 'test',
          sourceUrl: 'test',
          sourceName: 'test',
          readyInMinutes: 15,
          pricePerServing: 1,
          occasions: [],
          lowFodmap: false,
          license: 'test',
          ketogenic: false,
          instructions: 'test',
          healthScore: 12,
          glutenFree: false,
          gaps: 'test',
          extendedIngredients: new Set(),
          dishTypes: [],
          diets: [],
          dairyFree: false,
          cuisines: [],
          creditsText: 'test',
          cheap: false,
          analyzedInstructions: [{ steps: [] }],
          aggregateLikes: 1,
          winePairing: {
            pairedWines: [],
            pairingText: 'test',
            productMatches: new Set(),
          },
        });
      }),
    };
  });
  return {
    RecipesApi: MockInstance,
  };
});

// skipped till issue will be resolved
// https://github.com/vitest-dev/vitest/issues/5004
describe('Home tests', () => {
  const renderApp = () => {
    return render(
      <MemoryRouter initialEntries={['/']}>
        <RecipesApiProvider>
          <Home />
        </RecipesApiProvider>
      </MemoryRouter>
    );
  };
  it('should render home', async () => {
    const { asFragment, queryByTestId } = renderApp();

    await waitFor(() => {
      expect(queryByTestId('card')).toBeInTheDocument();
    });

    expect(asFragment()).toMatchSnapshot();
  });
});
