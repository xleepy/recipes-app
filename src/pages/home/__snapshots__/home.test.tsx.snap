// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Home tests > should render home 1`] = `
<DocumentFragment>
  <div>
    <input
      class="_search_4d1a47"
      data-testid="search"
      placeholder="Enter query"
      type="text"
      value="pasta"
    />
    <ul
      class="_card-container_13d6a5"
    >
      <div
        class="_card_c37972"
        data-testid="card"
        role="button"
        tabindex="0"
      >
        <img
          alt="preview"
          class="_image_c37972"
          loading="lazy"
          src="none"
        />
        <p>
          test
        </p>
      </div>
    </ul>
  </div>
</DocumentFragment>
`;
